File: apps/web/src/lib/community-v2/publish-client.ts
// The v2 publication flow on the Web (spec §31–§34), in one place:
//
//   1. post controls: the device's active root (created silently the first
//      time — no consent friction — unless a server vault exists that this
//      device has not restored yet);
//   2. eligibility: blind a token under the issuer key, ask Core (signed in),
//      finalize → a token Core cannot recognise, bound to the scope/context it
//      verified;
//   3. envelope: signed with the school/year posting key; a fresh post nonce
//      gives this post its own control key;
//   4. check → lane (+ pass) → publish, both to Community with credentials
//      omitted.
//
// Mine, revoke, reactions and reports use the same keys with challenges.

import {
  blindToken,
  finalizeToken,
  fromBase64Url,
  importIssuerPublicKey,
  postControlKeyPair,
  postingKeyPair,
  randomBytes,
  reactionKeyPair,
  signStatement,
  toBase64Url,
  type CheckResponseV2,
  type CommunityScope,
  type EligibilityInfo,
  type EligibilityToken,
  type IssuerDescriptor,
  type MineExperience,
  type SignedPostEnvelopeV2,
} from "@honey/shared/community-v2";
import { api } from "../../api/client";
import { community } from "../../api/community";
import { apiCache } from "../useApi";
import { localPostControls, type UnlockedRoots } from "./local-store";
import { postControls } from "./post-controls";

export class PostControlsUnavailable extends Error {
  constructor(readonly reason: "restore_needed" | "unsupported") {
    super(reason);
    this.name = "PostControlsUnavailable";
  }
}

export interface PublishTarget {
  lessonId?: string;
  entityKey?: string;
}

interface Session {
  scope: CommunityScope;
  issuer: IssuerDescriptor;
  issuerKey: CryptoKey;
}

let sessionPromise: Promise<Session> | null = null;

/** Scope (canonical exposure) and issuer descriptor, fetched once per session. */
export function communitySession(): Promise<Session> {
  sessionPromise ??= (async () => {
    const [scope, issuer] = await Promise.all([api.communityScope(), api.communityIssuer()]);
    return { scope, issuer, issuerKey: await importIssuerPublicKey(issuer.publicKey) };
  })().catch((err) => {
    sessionPromise = null;
    throw err;
  });
  return sessionPromise;
}

export function resetCommunitySession(): void {
  sessionPromise = null;
}

/** The active root for publishing; created silently when none exists anywhere. */
export async function ensureRoots(account: string): Promise<UnlockedRoots> {
  const status = await postControls.status(account);
  switch (status.kind) {
    case "ready":
    case "local_only":
      return status.roots;
    case "none":
      return postControls.create(account);
    case "restore_needed":
      throw new PostControlsUnavailable("restore_needed");
    case "unsupported":
      throw new PostControlsUnavailable("unsupported");
  }
}

/**
 * Blind issuance. Step 1 asks the issuer which metadata it would bind for the
 * target (standing checked, nothing signed or counted); the client blinds
 * under exactly that; step 2 is the one counted signing round.
 */
export async function obtainToken(target: PublishTarget | { schoolMember: true }): Promise<{ token: EligibilityToken; info: EligibilityInfo }> {
  const session = await communitySession();
  const request = "schoolMember" in target ? { schoolMember: true } : target;
  const stated = await api.communityEligibilityInfo(request);
  const blinded = await blindToken(session.issuerKey, stated.info);
  const issued = await api.communityEligibility({ ...request, blindedMessage: toBase64Url(blinded.blindedMessage) });
  const token = await finalizeToken(session.issuerKey, issued.keyId, blinded, issued.info, fromBase64Url(issued.blindSignature));
  return { token, info: issued.info };
}

export interface PreparedPost {
  token: EligibilityToken;
  envelope: SignedPostEnvelopeV2;
  postSignature: string;
}

export async function preparePost(account: string, target: PublishTarget, body: string, rating: number | null): Promise<PreparedPost> {
  const roots = await ensureRoots(account);
  const { token, info } = await obtainToken(target);
  const epoch = { schoolId: info.schoolId, academicYear: info.academicYear };
  const posting = postingKeyPair(roots.active, epoch);
  const postNonce = randomBytes(32);
  const control = postControlKeyPair(roots.active, postNonce, epoch);
  const sep = info.scope.indexOf(":");
  const envelope: SignedPostEnvelopeV2 = {
    protocolVersion: 2,
    schoolId: info.schoolId,
    academicYear: info.academicYear,
    primaryEntity: { type: info.scope.slice(0, sep) as SignedPostEnvelopeV2["primaryEntity"]["type"], id: info.scope.slice(sep + 1) },
    contexts: { ...info.contexts },
    body: body.trim(),
    rating,
    postNonce: toBase64Url(postNonce),
    postingPublicKey: toBase64Url(posting.publicKey),
    controlPublicKey: toBase64Url(control.publicKey),
    clientNonce: toBase64Url(randomBytes(12)),
  };
  return { token, envelope, postSignature: toBase64Url(signStatement(posting.privateKey, envelope as never)) };
}

export function checkPost(prepared: PreparedPost, cooldownTicket?: string): Promise<CheckResponseV2> {
  return community.check({ ...prepared, ...(cooldownTicket ? { cooldownTicket } : {}) });
}

export async function publishPost(prepared: PreparedPost, pass: string): Promise<{ experienceId: string }> {
  const res = await community.publish({ ...prepared, pass });
  apiCache.invalidate("community");
  return { experienceId: res.experienceId };
}

// ---- mine / revoke ----------------------------------------------------------

export interface OwnedPost extends MineExperience {
  rootId: string;
  academicYear: string;
}

/** Every post any root controls, across every school/year epoch known. */
export async function listOwnedPosts(account: string): Promise<OwnedPost[]> {
  const roots = await localPostControls.unlock(account);
  if (!roots) return [];
  const epochs = await postControls.epochs(account, roots);
  const out: OwnedPost[] = [];
  for (const root of roots.roots) {
    for (const epoch of epochs) {
      const posting = postingKeyPair(root.secret, epoch);
      const ch = await community.mineChallenge();
      const statement = { purpose: "honey/v2/mine" as const, schoolId: epoch.schoolId, academicYear: epoch.academicYear, challenge: ch.challenge, expiresAt: ch.expiresAt };
      const res = await community.mine({ statement, postingPublicKey: toBase64Url(posting.publicKey), signature: toBase64Url(signStatement(posting.privateKey, statement)) });
      for (const e of res.experiences) out.push({ ...e, rootId: root.rootId, academicYear: epoch.academicYear });
    }
  }
  return out.sort((a, b) => b.createdAt - a.createdAt);
}

export async function revokePost(account: string, post: OwnedPost, schoolId: string): Promise<void> {
  const roots = await localPostControls.unlock(account);
  const root = roots?.roots.find((r) => r.rootId === post.rootId);
  if (!root) throw new Error("the root that controls this post is not on this device");
  const control = postControlKeyPair(root.secret, fromBase64Url(post.postNonce), { schoolId, academicYear: post.academicYear });
  if (toBase64Url(control.publicKey) !== post.controlPublicKey) throw new Error("control key mismatch");
  const ch = await community.revokeChallenge(post.id);
  const statement = { purpose: "honey/v2/revoke" as const, experienceId: post.id, challenge: ch.challenge, expiresAt: ch.expiresAt };
  await community.revoke(post.id, { statement, signature: toBase64Url(signStatement(control.privateKey, statement)) });
  apiCache.invalidate("community");
}

// ---- reactions / reports -------------------------------------------------------

const REGISTERED_KEY = "honey.reactor.registered";

function registeredEpochs(): Set<string> {
  try {
    return new Set(JSON.parse(localStorage.getItem(REGISTERED_KEY) ?? "[]") as string[]);
  } catch {
    return new Set();
  }
}

async function reactorFor(account: string): Promise<{ privateKey: Uint8Array; publicKey: string; schoolId: string; academicYear: string }> {
  const roots = await ensureRoots(account);
  const session = await communitySession();
  const epoch = { schoolId: session.scope.schoolId, academicYear: session.scope.academicYear };
  const key = reactionKeyPair(roots.active, epoch);
  const publicKey = toBase64Url(key.publicKey);
  const mark = `${epoch.schoolId}\0${epoch.academicYear}\0${publicKey}`;
  const known = registeredEpochs();
  if (!known.has(mark)) {
    // Membership token → the reactor key is registered once per school/year.
    const { token } = await obtainToken({ schoolMember: true });
    const statement = { purpose: "honey/v2/register-reactor" as const, schoolId: epoch.schoolId, academicYear: epoch.academicYear, reactionPublicKey: publicKey };
    await community.registerReactor({ token, statement, signature: toBase64Url(signStatement(key.privateKey, statement)) });
    known.add(mark);
    try {
      localStorage.setItem(REGISTERED_KEY, JSON.stringify([...known]));
    } catch {
      /* the next reaction registers again; the server is idempotent */
    }
  }
  return { privateKey: key.privateKey, publicKey, ...epoch };
}

export async function reactToPost(account: string, experienceId: string, value: 1 | -1 | 0) {
  const r = await reactorFor(account);
  const statement = { purpose: "honey/v2/react" as const, schoolId: r.schoolId, academicYear: r.academicYear, experienceId, value, nonce: toBase64Url(randomBytes(12)) };
  return community.react(experienceId, { statement, reactionPublicKey: r.publicKey, signature: toBase64Url(signStatement(r.privateKey, statement)) });
}

export async function reportPost(account: string, experienceId: string, category: string) {
  const r = await reactorFor(account);
  const statement = { purpose: "honey/v2/report" as const, schoolId: r.schoolId, academicYear: r.academicYear, experienceId, category, nonce: toBase64Url(randomBytes(12)) };
  return community.report(experienceId, { statement, reactionPublicKey: r.publicKey, signature: toBase64Url(signStatement(r.privateKey, statement)) });
}

// The viewer's own reactions, remembered on this device (Community is not
// told who reads: the feed carries no per-viewer state).
const MY_REACTIONS_KEY = "honey.reactions.mine";
export const myReactions = {
  get(id: string): 1 | -1 | 0 {
    try {
      const v = (JSON.parse(localStorage.getItem(MY_REACTIONS_KEY) ?? "{}") as Record<string, number>)[id];
      return v === 1 || v === -1 ? v : 0;
    } catch {
      return 0;
    }
  },
  set(id: string, value: 1 | -1 | 0): void {
    try {
      const all = JSON.parse(localStorage.getItem(MY_REACTIONS_KEY) ?? "{}") as Record<string, number>;
      if (value === 0) delete all[id];
      else all[id] = value;
      localStorage.setItem(MY_REACTIONS_KEY, JSON.stringify(all));
    } catch {
      /* best effort */
    }
  },
};

File: ios-web-port/HOneyCore/Sources/HOneyCore/CommunityV2/PublishClient.swift
// The v2 publication flow on the iPhone (spec §31–§34; Web:
// lib/community-v2/publish-client.ts), in one place:
//
//   1. post controls: the device's active root (created silently at the
//      first share unless a server vault exists that this device has not
//      restored yet);
//   2. eligibility: ask the issuer which metadata it would bind (uncounted),
//      blind a token under exactly that, one counted signing round, finalize
//      → a token Core cannot recognise, bound to the scope it verified;
//   3. envelope: signed with the school/year posting key; a fresh post nonce
//      gives this post its own control key;
//   4. check → lane (+ pass) → publish, both to Community with no identity.
//
// Mine, revoke, reactions and reports use the same keys with challenges.

import Foundation

/// The issuer/scope endpoints of Core the flow needs (APIClient conforms).
public protocol IssuerAPI: Sendable {
    func communityIssuer() async throws -> IssuerDescriptor
    func communityScope() async throws -> CommunityScope
    func communityEligibilityInfo(_ target: EligibilityTarget) async throws -> EligibilityInfoResponse
    func communityEligibility(_ request: EligibilityRequest) async throws -> EligibilityIssued
}

/// Device memory of the viewer's own reactions and reactor registrations
/// (Community keeps no per-viewer state; the feed carries none).
public protocol ReactionMemory: Sendable {
    func myReaction(_ experienceId: String) -> Int
    func setMyReaction(_ experienceId: String, _ value: Int)
    func reactorRegistered(_ mark: String) -> Bool
    func setReactorRegistered(_ mark: String)
}

public struct PublishTarget: Sendable, Equatable, Hashable {
    public var lessonId: String?
    public var entityKey: String?
    public init(lessonId: String? = nil, entityKey: String? = nil) {
        self.lessonId = lessonId
        self.entityKey = entityKey
    }
    var eligibility: EligibilityTarget { EligibilityTarget(lessonId: lessonId, entityKey: entityKey) }
}

public struct PreparedPost: Sendable, Equatable {
    public var token: EligibilityToken
    public var envelope: SignedPostEnvelopeV2
    public var postSignature: String
}

public struct OwnedPost: Sendable, Equatable, Identifiable {
    public var experience: MineExperience
    public var rootId: String
    public var epoch: SchoolEpoch
    public var id: String { experience.id }
}

public enum PublishError: Error, Sendable, Equatable {
    case postControlsRestoreNeeded
    case rootNotOnThisDevice
    case controlKeyMismatch
}

public actor PublishClient {
    private let api: IssuerAPI
    private let community: CommunityAPIClient
    private let controls: PostControls
    private let memory: ReactionMemory
    private var session: (scope: CommunityScope, issuer: IssuerDescriptor, key: IssuerRSAPublicKey)?

    public init(api: IssuerAPI, community: CommunityAPIClient, controls: PostControls, memory: ReactionMemory) {
        self.api = api
        self.community = community
        self.controls = controls
        self.memory = memory
    }

    /// Scope (canonical exposure) and issuer descriptor, fetched once per session.
    public func communitySession() async throws -> (scope: CommunityScope, issuer: IssuerDescriptor, key: IssuerRSAPublicKey) {
        if let session { return session }
        async let scope = api.communityScope()
        async let issuer = api.communityIssuer()
        let s = try await scope
        let i = try await issuer
        let session = (s, i, try IssuerRSAPublicKey(descriptor: i))
        self.session = session
        return session
    }

    public func resetSession() { session = nil }

    public func exposure() async throws -> ExposureScope { try await communitySession().scope.exposure }

    private func roots(account: String) async throws -> UnlockedRoots {
        let epoch = try await communitySession().scope.epoch
        do {
            return try await controls.ensureRoots(account: account, epoch: epoch)
        } catch PostControlsError.restoreNeeded {
            throw PublishError.postControlsRestoreNeeded
        }
    }

    // MARK: eligibility (two rounds; only the second is counted)

    public func obtainToken(_ target: EligibilityTarget) async throws -> (token: EligibilityToken, info: EligibilityInfo) {
        let session = try await communitySession()
        let stated = try await api.communityEligibilityInfo(target)
        let info = try BlindToken.infoBytes(stated.info)
        let message = BlindToken.prepare(Data.random(32))
        let blinded = try BlindToken.blind(publicKey: session.key, message: message, info: info)
        let issued = try await api.communityEligibility(EligibilityRequest(target: target, blindedMessage: Base64URL.encode(blinded.blindedMessage)))
        guard let blindSig = Base64URL.decode(issued.blindSignature) else { throw BlindTokenError.unexpectedInputSize }
        let signature = try BlindToken.finalize(publicKey: session.key, blinded: blinded, blindSignature: blindSig)
        return (EligibilityToken(keyId: issued.keyId, info: issued.info, message: Base64URL.encode(message), signature: Base64URL.encode(signature)), issued.info)
    }

    // MARK: prepare · check · publish

    public func preparePost(account: String, target: PublishTarget, body: String, rating: Int?) async throws -> PreparedPost {
        let roots = try await roots(account: account)
        let (token, info) = try await obtainToken(target.eligibility)
        let epoch = info.epoch
        let posting = try V2Derivation.postingKeyPair(root: roots.active.secret, epoch: epoch)
        let postNonce = Data.random(32)
        let control = try V2Derivation.postControlKeyPair(root: roots.active.secret, postNonce: postNonce, epoch: epoch)
        let scope = info.scopeParts
        let envelope = SignedPostEnvelopeV2(
            schoolId: info.schoolId,
            academicYear: info.academicYear,
            primaryEntity: EnvelopeEntity(type: scope.type, id: scope.id),
            contexts: EnvelopeContexts(info.contexts),
            body: body.trimmingCharacters(in: .whitespacesAndNewlines),
            rating: rating,
            postNonce: Base64URL.encode(postNonce),
            postingPublicKey: Base64URL.encode(posting.publicKey),
            controlPublicKey: Base64URL.encode(control.publicKey),
            clientNonce: Base64URL.encode(.random(12))
        )
        let signature = try V2Derivation.signStatement(privateKey: posting.privateKey, envelope)
        return PreparedPost(token: token, envelope: envelope, postSignature: Base64URL.encode(signature))
    }

    public func check(_ prepared: PreparedPost, cooldownTicket: String?) async throws -> CheckResponseV2 {
        try await community.check(CheckRequestV2(token: prepared.token, envelope: prepared.envelope, postSignature: prepared.postSignature, cooldownTicket: cooldownTicket))
    }

    public func publish(_ prepared: PreparedPost, pass: String) async throws -> PublishResponseV2 {
        try await community.publish(PublishRequestV2(token: prepared.token, envelope: prepared.envelope, postSignature: prepared.postSignature, pass: pass))
    }

    // MARK: mine / revoke

    /// Every post any root controls, across every school/year epoch known.
    public func listOwnedPosts(account: String) async throws -> [OwnedPost] {
        guard let roots = try await controls.unlock(account: account) else { return [] }
        let epochs = await controls.epochs(account: account, roots: roots)
        var out: [OwnedPost] = []
        for root in roots.roots {
            for epoch in epochs {
                let posting = try V2Derivation.postingKeyPair(root: root.secret, epoch: epoch)
                let challenge = try await community.mineChallenge()
                let statement = MineStatement(schoolId: epoch.schoolId, academicYear: epoch.academicYear, challenge: challenge.challenge, expiresAt: challenge.expiresAt)
                let signature = try V2Derivation.signStatement(privateKey: posting.privateKey, statement)
                let res = try await community.mine(MineRequest(statement: statement, postingPublicKey: Base64URL.encode(posting.publicKey), signature: Base64URL.encode(signature)))
                for e in res.experiences { out.append(OwnedPost(experience: e, rootId: root.rootId, epoch: epoch)) }
            }
        }
        return out.sorted { $0.experience.createdAt > $1.experience.createdAt }
    }

    public func revoke(account: String, post: OwnedPost) async throws {
        guard let roots = try await controls.unlock(account: account), let root = roots.roots.first(where: { $0.rootId == post.rootId }) else {
            throw PublishError.rootNotOnThisDevice
        }
        guard let nonce = Base64URL.decode(post.experience.postNonce) else { throw PublishError.controlKeyMismatch }
        let control = try V2Derivation.postControlKeyPair(root: root.secret, postNonce: nonce, epoch: post.epoch)
        guard Base64URL.encode(control.publicKey) == post.experience.controlPublicKey else { throw PublishError.controlKeyMismatch }
        let challenge = try await community.revokeChallenge(experienceId: post.id)
        let statement = RevokeStatement(experienceId: post.id, challenge: challenge.challenge, expiresAt: challenge.expiresAt)
        let signature = try V2Derivation.signStatement(privateKey: control.privateKey, statement)
        _ = try await community.revoke(experienceId: post.id, RevokeRequest(statement: statement, signature: Base64URL.encode(signature)))
    }

    // MARK: reactions / reports

    private func reactor(account: String) async throws -> (privateKey: Data, publicKey: String, epoch: SchoolEpoch) {
        let roots = try await roots(account: account)
        let epoch = try await communitySession().scope.epoch
        let key = try V2Derivation.reactionKeyPair(root: roots.active.secret, epoch: epoch)
        let publicKey = Base64URL.encode(key.publicKey)
        let mark = "\(epoch.schoolId)\0\(epoch.academicYear)\0\(publicKey)"
        if !memory.reactorRegistered(mark) {
            // Membership token → the reactor key is registered once per school/year (the server is idempotent).
            let (token, _) = try await obtainToken(.member)
            let statement = RegisterReactorStatement(schoolId: epoch.schoolId, academicYear: epoch.academicYear, reactionPublicKey: publicKey)
            let signature = try V2Derivation.signStatement(privateKey: key.privateKey, statement)
            _ = try await community.registerReactor(RegisterReactorRequest(token: token, statement: statement, signature: Base64URL.encode(signature)))
            memory.setReactorRegistered(mark)
        }
        return (key.privateKey, publicKey, epoch)
    }

    public func react(account: String, experienceId: String, value: Int) async throws -> ReactResponseV2 {
        let r = try await reactor(account: account)
        let statement = ReactStatement(schoolId: r.epoch.schoolId, academicYear: r.epoch.academicYear, experienceId: experienceId, value: value, nonce: Base64URL.encode(.random(12)))
        let signature = try V2Derivation.signStatement(privateKey: r.privateKey, statement)
        let res = try await community.react(experienceId: experienceId, ReactRequestV2(statement: statement, reactionPublicKey: r.publicKey, signature: Base64URL.encode(signature)))
        memory.setMyReaction(experienceId, res.value)
        return res
    }

    public func report(account: String, experienceId: String, category: ReportCategory) async throws {
        let r = try await reactor(account: account)
        let statement = ReportStatement(schoolId: r.epoch.schoolId, academicYear: r.epoch.academicYear, experienceId: experienceId, category: category.rawValue, nonce: Base64URL.encode(.random(12)))
        let signature = try V2Derivation.signStatement(privateKey: r.privateKey, statement)
        _ = try await community.report(experienceId: experienceId, ReportRequestV2(statement: statement, reactionPublicKey: r.publicKey, signature: Base64URL.encode(signature)))
    }
}
